### PR TITLE
Fix threads framework

### PR DESCRIPTION
**Pull request by shintaro-iwasaki**
This PR addresses most of the problems newly found except the following items that I think are still discussed:

- TSD management 
- Definitions of `REQUEST_PENDING` and `REQUEST_COMPLETED`
- Some cleanups.

The head repository is 
```
https://github.com/shintaro-iwasaki/ompi-1
branch: topic/thread_framework2
```

Could you please squash this branch so that I can resolve some of his comments.

### CODE REVIEW

**Comment by hppritcha** [comment]
I'm finding some problems here. Will do a review.

File: opal/mca/threads/pthreads/Makefile.am [hunk 1]
```diff
@@ -30,3 +30,5 @@ libmca_threads_pthreads_la_SOURCES = \
     threads_pthreads_tsd.h \
     threads_pthreads_wait_sync.c \
     threads_pthreads_wait_sync.h
+
+AM_LDFLAGS = -labt
```
Comment (hppritcha) [inline opal/mca/threads/pthreads/Makefile.am]: I don't think this is right.  Why add -labt for pthreads specific Makefile.am?

**Comment by shintaro-iwasaki** [comment]
Thanks!

File: opal/mca/threads/pthreads/threads_pthreads_mutex.h [hunk 2]
```diff
@@ -115,8 +115,6 @@ OPAL_DECLSPEC OBJ_CLASS_DECLARATION(opal_recursive_mutex_t);
     }
 #endif
 
-#endif
```
Comment (hppritcha) [inline opal/mca/threads/pthreads/threads_pthreads_mutex.h]: why did you remove this endif?  it looks correct to me.  It closes the defined(OPAL_PTHREAD_RECURSIVE_MUTEX_INITIALIZER) no?

Comment (shintaro-iwasaki) [inline opal/mca/threads/pthreads/threads_pthreads_mutex.h]: Here I should not remove this `#endif`. I have to check qthreads and argobots too.

File: opal/mca/threads/argobots/threads_argobots.h [hunk 3]
```diff
@@ -30,7 +30,7 @@
 
 static inline void opal_threads_argobots_ensure_init(void)
 {
-    if (ABT_initialized() != 0) {
+    if (ABT_initialized() != ABT_SUCCESS) {
```
Comment (hppritcha) [inline opal/mca/threads/argobots/threads_argobots.h]: would you mind reversing the check to have ABT_SUCCESS != ABT_initialize()?  I know it looks weird but that's the "openmpi" coding style.

**Review by hppritcha — CHANGES_REQUESTED** [review]

File: opal/mca/threads/pthreads/Makefile.am (hunk 1 above)
Comment (shintaro-iwasaki) [reply opal/mca/threads/pthreads/Makefile.am]: You are right. It is wrong. I will fix it soon.

File: opal/mca/threads/argobots/threads_argobots.h (hunk 3 above)
Comment (shintaro-iwasaki) [reply opal/mca/threads/argobots/threads_argobots.h]: Thank you. I will follow the Open MPI style. I will examine other checks.

File: opal/mca/threads/pthreads/Makefile.am (hunk 1 above)
Comment (shintaro-iwasaki) [reply opal/mca/threads/pthreads/Makefile.am]: I wrongly put it in `pthreads/Makefile.am`. This `AM_LDFLAGS` was moved to `argobots/Makefile.am`.

File: opal/mca/threads/pthreads/threads_pthreads_mutex.h (hunk 2 above)
Comment (shintaro-iwasaki) [reply opal/mca/threads/pthreads/threads_pthreads_mutex.h]: This change should have been done to `qthreads/threads_qthreads_mutex.h` instead. Fixed.

File: opal/mca/threads/argobots/threads_argobots.h (hunk 3 above)
Comment (shintaro-iwasaki) [reply opal/mca/threads/argobots/threads_argobots.h]: I found some did not follow.
```
if (NULL == t->t_run || t->t_handle 1= (pthread_t)-1) {
to
if (NULL == t->t_run || (pthread_t)-1 != t->t_handle) {
```
I fixed them as far as I could find.

**Comment by shintaro-iwasaki** [comment]
I fixed them. Could you please look at this PR again?

**Review by hppritcha — APPROVED** [review]
